File: libraries/CSharp/Services/ExtensionService.cs
using System.Net.Http.Json;
using GPMLoginGlobal.Models;

namespace GPMLoginGlobal.Services;

/// <summary>
/// Provides methods for managing browser extensions via the GPMLogin Global API.
/// </summary>
public sealed class ExtensionService(HttpClient http)
{
    /// <summary>
    /// Retrieves all installed extensions.
    /// </summary>
    /// <param name="cancellationToken">Token to cancel the request.</param>
    /// <returns>A read-only list of <see cref="Extension"/> objects.</returns>
    /// <exception cref="InvalidOperationException">Thrown when the API returns <c>success: false</c>.</exception>
    public async Task<IReadOnlyList<Extension>> GetAllAsync(CancellationToken cancellationToken = default)
    {
        ApiResponse<IReadOnlyList<Extension>>? response =
            await http.GetFromJsonAsync<ApiResponse<IReadOnlyList<Extension>>>("extensions", cancellationToken);

        EnsureSuccess(response);
        return response!.Data!;
    }

    /// <summary>
    /// Enables or disables a specific extension.
    /// </summary>
    /// <param name="id">The extension identifier.</param>
    /// <param name="active"><see langword="true"/> to enable the extension; <see langword="false"/> to disable it.</param>
    /// <param name="cancellationToken">Token to cancel the request.</param>
    /// <exception cref="InvalidOperationException">Thrown when the API returns <c>success: false</c>.</exception>
    public async Task UpdateStateAsync(string id, bool active, CancellationToken cancellationToken = default)
    {
        string activeStr = active ? "true" : "false";

        ApiResponse<object>? response =
            await http.GetFromJsonAsync<ApiResponse<object>>(
                $"extensions/update-state/{id}?active={activeStr}",
                cancellationToken);

        EnsureSuccess(response);
    }

    // -------------------------------------------------------------------------
    // Helpers
    // -------------------------------------------------------------------------

    private static void EnsureSuccess<T>(ApiResponse<T>? response)
    {
        if (response is null)
            throw new InvalidOperationException("The API returned an empty response.");

        if (!response.Success)
            throw new InvalidOperationException($"GPMLogin API error: {response.Message}");
    }
}

File: libraries/Js/src/services/ExtensionService.js
/**
 * @file ExtensionService.js
 * Service for managing browser extensions via the GPMLogin Global API.
 */

/**
 * @param {object} response
 * @throws {Error}
 */
function ensureSuccess(response) {
  if (!response.success) {
    throw new Error(`GPMLogin API error: ${response.message}`);
  }
}

/**
 * Provides operations for listing and toggling browser extensions.
 */
export class ExtensionService {
  /**
   * @param {{ get: Function, post: Function }} http - Shared HTTP helper.
   */
  constructor(http) {
    this._http = http;
  }

  /**
   * Retrieves all installed browser extensions.
   *
   * @returns {Promise<Array<object>>} Array of extension objects, each with
   *   `id`, `name`, and `active` fields.
   * @throws {Error} On API failure.
   *
   * @example
   * const extensions = await client.extensions.getAll();
   * extensions.forEach(ext => console.log(ext.name, ext.active));
   */
  async getAll() {
    const response = await this._http.get('extensions');
    ensureSuccess(response);
    return response.data;
  }

  /**
   * Enables or disables a specific extension.
   *
   * @param {string} id - The extension identifier.
   * @param {boolean} active - `true` to enable, `false` to disable.
   * @returns {Promise<void>}
   * @throws {Error} On API failure.
   *
   * @example
   * await client.extensions.updateState('ext-id-123', true);
   */
  async updateState(id, active) {
    const response = await this._http.get(
      `extensions/update-state/${id}?active=${active ? 'true' : 'false'}`
    );
    ensureSuccess(response);
  }
}
